torch/distributions/categorical.py

```diff
@@ -47,7 +47,9 @@ def __init__(self, probs=None, logits=None, validate_args=None):
         if probs is not None:
             if probs.dim() < 1:
                 raise ValueError("`probs` parameter must be at least one-dimensional.")
-            self.probs = probs / probs.sum(-1, keepdim=True)
+            sum_probs = probs.detach().cpu().numpy().sum(-1, keepdims=True)
```

Comment:
This places a dependency on NumPy unnecessarily.

Reply:
@vishwakftw `.numpy()` will returns a numpy array, but we do not explicitly require `import numpy` in this file. Is this what you mean ?

Reply:
`numpy()` doesn't work unless you have a copy of PyTorch which was compiled with Numpy support. We don't want to assume that this is the case in core code.

Reply:
@ezyang Thanks for the comment. In this case, I would think to remove the 're-normalize' operation and let user to take responsibility for the validity of probabilities. Because with this line, it is really really slow in certain cases e.g. I am training policy with ES to solve RL tasks, the discrete (`Categorical`) environment is extremely slow compared with continuous tasks (`Normal`), because we have to create distribution for each time step. In addition, before creating categorical distribution, the most common use cases are either to take a softmax and feed probabilities in or feed the logits in. Thus, I would vote for removing this line of code as the violation of validity is rare but the degrading of speed is very high. What do you think ?

Reply:
I did benchmark the operation: sum along the last dim and add, and it seems to be slower on NumPy actually.
With PyTorch (cpu):
`33.2 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`
With PyTorch (gpu, 1080Ti):
`3.74 ms ± 28.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`
With NumPy:
`189 ms ± 8.33 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`

And the `detach` is breaking autograd tests.

Reply:
@vishwakftw I made the following test case

```
from time import time
import numpy as np
import torch
import torch.nn.functional as F

t = time()
for _ in range(100):
    p = F.softmax(torch.randn(1, 32), -1)
    p = p/p.sum(-1, keepdim=True)
print(time() - t)

t = time()
for _ in range(100):
    p = F.softmax(torch.randn(1, 32), -1)
    
    p = p/torch.from_numpy(p.detach().numpy().sum(-1, keepdims=True)).float()
print(time() - t)
```
the Numpy alternative is much faster.

But anyway as @ezyang pointed out above, numpy should not be considered here. 

The original motivation is in #11747 . Do you have some idea how to reimplement it without using `.sum()` function ?

Reply:
Can't reproduce those times btw, running your code gives me 
`0.007228851318359375
0.007979869842529297` with pytorch 0.4.1, on google colab no gpu

Reply:
@alicanb @vishwakftw Indeed. The main use case I am referring to is using ES to evolve a small policy to solve sequential problems (RL), which requires a lot of samples each over a long horizon, this rises of the need being able to create thousands of 'small' categorical distributions efficiently. On the ES code I tested on, with same setting, it becomes extremely slow compared with other distributions like `Normal` for continuous control tasks.